### ISSUE TITLE
training time is too long for data process

### ISSUE THREAD

**Issue by liyuanyaun** [weiliu89/caffe#810]
i use ssd_pascal.py for training, and my training log is as following:
I1207 14:33:58.859539 17469 solver.cpp:259]     Train net output #0: mbox_loss = 12.2819 (* 1 = 12.2819 loss)
I1207 14:34:00.304049 17469 sgd_solver.cpp:138] Iteration 0, lr = 0.01
I1207 14:34:02.271486 17490 annotated_data_layer.cpp:305] Prefetch batch: 5348 ms.
I1207 14:34:02.271548 17490 annotated_data_layer.cpp:306]      Read time: 0.22 ms.
I1207 14:34:02.271555 17490 annotated_data_layer.cpp:307] Transform time: 5288.89 ms.
I1207 14:34:04.866214 17502 annotated_data_layer.cpp:305] Prefetch batch: 6003 ms.
I1207 14:34:04.866261 17502 annotated_data_layer.cpp:306]      Read time: 0.282 ms.
I1207 14:34:04.866281 17502 annotated_data_layer.cpp:307] Transform time: 5951.13 ms.
I1207 14:34:09.448494 17490 annotated_data_layer.cpp:305] Prefetch batch: 7125 ms.
I1207 14:34:09.448555 17490 annotated_data_layer.cpp:306]      Read time: 0.183 ms.
I1207 14:34:09.448562 17490 annotated_data_layer.cpp:307] Transform time: 7042.11 ms.
I1207 14:34:11.907380 17502 annotated_data_layer.cpp:305] Prefetch batch: 7003 ms.
I1207 14:34:11.907428 17502 annotated_data_layer.cpp:306]      Read time: 0.276 ms.
I1207 14:34:11.907434 17502 annotated_data_layer.cpp:307] Transform time: 6967.15 ms.
I1207 14:34:15.350066 17490 annotated_data_layer.cpp:305] Prefetch batch: 5867 ms.
I1207 14:34:15.350247 17490 annotated_data_layer.cpp:306]      Read time: 0.179 ms.
I1207 14:34:15.350256 17490 annotated_data_layer.cpp:307] Transform time: 5853.3 ms.
I1207 14:34:16.723649 17502 annotated_data_layer.cpp:305] Prefetch batch: 4767 ms.
I1207 14:34:16.723709 17502 annotated_data_layer.cpp:306]      Read time: 0.343 ms.
I1207 14:34:16.723716 17502 annotated_data_layer.cpp:307] Transform time: 4760.15 ms.
I1207 14:34:21.469477 17490 annotated_data_layer.cpp:305] Prefetch batch: 6082 ms.
I1207 14:34:21.469527 17490 annotated_data_layer.cpp:306]      Read time: 0.128 ms.
I1207 14:34:21.469549 17490 annotated_data_layer.cpp:307] Transform time: 6073.1 ms.
I1207 14:34:23.284888 17502 annotated_data_layer.cpp:305] Prefetch batch: 6522 ms.
I1207 14:34:23.284932 17502 annotated_data_layer.cpp:306]      Read time: 0.302 ms.
I1207 14:34:23.284939 17502 annotated_data_layer.cpp:307] Transform time: 6509.66 ms.
I1207 14:34:26.264127 17490 annotated_data_layer.cpp:305] Prefetch batch: 4758 ms.
I1207 14:34:26.264187 17490 annotated_data_layer.cpp:306]      Read time: 0.134 ms.
I1207 14:34:26.264194 17490 annotated_data_layer.cpp:307] Transform time: 4748.85 ms.
I1207 14:34:30.159955 17502 annotated_data_layer.cpp:305] Prefetch batch: 6817 ms.
I1207 14:34:30.160001 17502 annotated_data_layer.cpp:306]      Read time: 0.283 ms.
I1207 14:34:30.160023 17502 annotated_data_layer.cpp:307] Transform time: 6807.06 ms.
I1207 14:34:31.422053 17490 annotated_data_layer.cpp:305] Prefetch batch: 5120 ms.
I1207 14:34:31.422101 17490 annotated_data_layer.cpp:306]      Read time: 0.142 ms.
I1207 14:34:31.422123 17490 annotated_data_layer.cpp:307] Transform time: 5110.52 ms.
I1207 14:34:35.778383 17502 annotated_data_layer.cpp:305] Prefetch batch: 5582 ms.
I1207 14:34:35.778442 17502 annotated_data_layer.cpp:306]      Read time: 0.258 ms.
I1207 14:34:35.778450 17502 annotated_data_layer.cpp:307] Transform time: 5573.74 ms.
I1207 14:34:37.514938 17490 annotated_data_layer.cpp:305] Prefetch batch: 6058 ms.
I1207 14:34:37.515003 17490 annotated_data_layer.cpp:306]      Read time: 0.139 ms.
I1207 14:34:37.515010 17490 annotated_data_layer.cpp:307] Transform time: 6049.4 ms.
I1207 14:34:41.990942 17490 annotated_data_layer.cpp:305] Prefetch batch: 4439 ms.
I1207 14:34:41.990988 17490 annotated_data_layer.cpp:306]      Read time: 0.133 ms.
I1207 14:34:41.991010 17490 annotated_data_layer.cpp:307] Transform time: 4428.52 ms.
I1207 14:34:43.524531 17502 annotated_data_layer.cpp:305] Prefetch batch: 7708 ms.
I1207 14:34:43.524616 17502 annotated_data_layer.cpp:306]      Read time: 0.351 ms.
I1207 14:34:43.524626 17502 annotated_data_layer.cpp:307] Transform time: 7698.38 ms.
I1207 14:34:48.599135 17490 annotated_data_layer.cpp:305] Prefetch batch: 6571 ms.
I1207 14:34:48.599457 17490 annotated_data_layer.cpp:306]      Read time: 0.152 ms.
I1207 14:34:48.599469 17490 annotated_data_layer.cpp:307] Transform time: 6559.94 ms.
I1207 14:34:48.740897 17502 annotated_data_layer.cpp:305] Prefetch batch: 5156 ms.
I1207 14:34:48.740958 17502 annotated_data_layer.cpp:306]      Read time: 0.289 ms.
I1207 14:34:48.740965 17502 annotated_data_layer.cpp:307] Transform time: 5145.92 ms.
I1207 14:34:52.994175 17502 annotated_data_layer.cpp:305] Prefetch batch: 4219 ms.
I1207 14:34:52.994235 17502 annotated_data_layer.cpp:306]      Read time: 0.246 ms.
I1207 14:34:52.994242 17502 annotated_data_layer.cpp:307] Transform time: 4209.21 ms.
I1207 14:36:41.661922 17502 annotated_data_layer.cpp:305] Prefetch batch: 4215 ms.
I1207 14:36:41.662202 17502 annotated_data_layer.cpp:306]      Read time: 0.189 ms.
I1207 14:36:41.662214 17502 annotated_data_layer.cpp:307] Transform time: 4187.27 ms.
I1207 14:36:43.763808 17490 annotated_data_layer.cpp:305] Prefetch batch: 6318 ms.
I1207 14:36:43.763870 17490 annotated_data_layer.cpp:306]      Read time: 0.332 ms.
I1207 14:36:43.763877 17490 annotated_data_layer.cpp:307] Transform time: 6296.57 ms.
I1207 14:36:47.775596 17469 solver.cpp:243] Iteration 10, loss = 12.4834
I1207 14:36:47.775684 17469 solver.cpp:259]     Train net output #0: mbox_loss = 7.75353 (* 1 = 7.75353 loss)
I1207 14:36:48.802582 17490 annotated_data_layer.cpp:305] Prefetch batch: 5004 ms.
I1207 14:36:48.802649 17490 annotated_data_layer.cpp:306]      Read time: 0.31 ms.
I1207 14:36:48.802671 17490 annotated_data_layer.cpp:307] Transform time: 4996.74 ms.
I1207 14:36:49.823837 17502 annotated_data_layer.cpp:305] Prefetch batch: 8124 ms.
I1207 14:36:49.823897 17502 annotated_data_layer.cpp:306]      Read time: 0.218 ms.
I1207 14:36:49.823904 17502 annotated_data_layer.cpp:307] Transform time: 8111.39 ms.
I1207 14:36:55.350520 17490 annotated_data_layer.cpp:305] Prefetch batch: 6511 ms.
I1207 14:36:55.350569 17490 annotated_data_layer.cpp:306]      Read time: 0.225 ms.
I1207 14:36:55.350594 17490 annotated_data_layer.cpp:307] Transform time: 6502.18 ms.
I1207 14:36:58.652791 17502 annotated_data_layer.cpp:305] Prefetch batch: 8793 ms.
I1207 14:36:58.652858 17502 annotated_data_layer.cpp:306]      Read time: 0.354 ms.
I1207 14:36:58.652866 17502 annotated_data_layer.cpp:307] Transform time: 8782.72 ms.
I1207 14:36:58.773773 17469 sgd_solver.cpp:138] Iteration 10, lr = 0.01
I1207 14:37:00.429812 17490 annotated_data_layer.cpp:305] Prefetch batch: 5043 ms.
I1207 14:37:00.429872 17490 annotated_data_layer.cpp:306]      Read time: 0.165 ms.
I1207 14:37:00.429878 17490 annotated_data_layer.cpp:307] Transform time: 4997.51 ms.
I1207 14:37:34.215801 17490 annotated_data_layer.cpp:305] Prefetch batch: 33748 ms.
I1207 14:37:34.215988 17490 annotated_data_layer.cpp:306]      Read time: 0.169 ms.
I1207 14:37:34.216013 17490 annotated_data_layer.cpp:307] Transform time: 33738.4 ms.
I1207 14:37:36.687965 17502 annotated_data_layer.cpp:305] Prefetch batch: 37984 ms.
I1207 14:37:36.688030 17502 annotated_data_layer.cpp:306]      Read time: 0.329 ms.
I1207 14:37:36.688035 17502 annotated_data_layer.cpp:307] Transform time: 37968.7 ms.
I1207 14:37:41.320627 17490 annotated_data_layer.cpp:305] Prefetch batch: 7069 ms.
I1207 14:37:41.320691 17490 annotated_data_layer.cpp:306]      Read time: 0.139 ms.
I1207 14:37:41.320698 17490 annotated_data_layer.cpp:307] Transform time: 7052.02 ms.
I1207 14:37:42.747607 17502 annotated_data_layer.cpp:305] Prefetch batch: 6026 ms.
I1207 14:37:42.747668 17502 annotated_data_layer.cpp:306]      Read time: 0.255 ms.
I1207 14:37:42.747675 17502 annotated_data_layer.cpp:307] Transform time: 6017.05 ms.
I1207 14:37:45.411252 17490 annotated_data_layer.cpp:305] Prefetch batch: 4054 ms.
I1207 14:37:45.411298 17490 annotated_data_layer.cpp:306]      Read time: 0.122 ms.
I1207 14:37:45.411319 17490 annotated_data_layer.cpp:307] Transform time: 4045.31 ms.
I1207 14:37:48.783057 17502 annotated_data_layer.cpp:305] Prefetch batch: 6001 ms.
I1207 14:37:48.783118 17502 annotated_data_layer.cpp:306]      Read time: 0.277 ms.
I1207 14:37:48.783124 17502 annotated_data_layer.cpp:307] Transform time: 5984.86 ms.
I1207 14:37:51.740692 17490 annotated_data_layer.cpp:305] Prefetch batch: 6292 ms.
I1207 14:37:51.740759 17490 annotated_data_layer.cpp:306]      Read time: 0.131 ms.
I1207 14:37:51.740766 17490 annotated_data_layer.cpp:307] Transform time: 6275.91 ms.
I1207 14:37:54.933676 17502 annotated_data_layer.cpp:305] Prefetch batch: 6116 ms.
I1207 14:37:54.933738 17502 annotated_data_layer.cpp:306]      Read time: 0.251 ms.
I1207 14:37:54.933745 17502 annotated_data_layer.cpp:307] Transform time: 6108.82 ms.
I1207 14:38:12.287582 17490 annotated_data_layer.cpp:305] Prefetch batch: 20494 ms.
I1207 14:38:12.287906 17490 annotated_data_layer.cpp:306]      Read time: 0.129 ms.
I1207 14:38:12.287926 17490 annotated_data_layer.cpp:307] Transform time: 20487.4 ms.
I1207 14:38:16.074512 17490 annotated_data_layer.cpp:305] Prefetch batch: 3748 ms.
I1207 14:38:16.074574 17490 annotated_data_layer.cpp:306]      Read time: 0.208 ms.
I1207 14:38:16.074581 17490 annotated_data_layer.cpp:307] Transform time: 3725.63 ms.
I1207 14:38:30.467836 17502 annotated_data_layer.cpp:305] Prefetch batch: 35500 ms.
I1207 14:38:30.467916 17502 annotated_data_layer.cpp:306]      Read time: 0.305 ms.
I1207 14:38:30.467926 17502 annotated_data_layer.cpp:307] Transform time: 35489.5 ms.
I1207 14:38:39.745079 17502 annotated_data_layer.cpp:305] Prefetch batch: 9243 ms.
I1207 14:38:39.745162 17502 annotated_data_layer.cpp:306]      Read time: 0.255 ms.
I1207 14:38:39.745170 17502 annotated_data_layer.cpp:307] Transform time: 9226.19 ms.
I1207 14:39:03.534462 17490 annotated_data_layer.cpp:305] Prefetch batch: 47422 ms.
I1207 14:39:03.534757 17490 annotated_data_layer.cpp:306]      Read time: 0.208 ms.
I1207 14:39:03.534768 17490 annotated_data_layer.cpp:307] Transform time: 47396.9 ms.
I1207 14:39:15.196792 17502 annotated_data_layer.cpp:305] Prefetch batch: 35417 ms.
I1207 14:39:15.196861 17502 annotated_data_layer.cpp:306]      Read time: 0.328 ms.
I1207 14:39:15.196868 17502 annotated_data_layer.cpp:307] Transform time: 35392.2 ms.
I1207 14:39:34.537024 17502 annotated_data_layer.cpp:305] Prefetch batch: 19306 ms.
I1207 14:39:34.578469 17502 annotated_data_layer.cpp:306]      Read time: 0.221 ms.
I1207 14:39:34.578526 17502 annotated_data_layer.cpp:307] Transform time: 19292.1 ms.
I1207 14:39:37.243183 17490 annotated_data_layer.cpp:305] Prefetch batch: 22988 ms.
I1207 14:39:37.243268 17490 annotated_data_layer.cpp:306]      Read time: 0.182 ms.
I1207 14:39:37.243283 17490 annotated_data_layer.cpp:307] Transform time: 22949.4 ms.
I1207 14:39:45.809635 17490 annotated_data_layer.cpp:305] Prefetch batch: 8166 ms.
I1207 14:39:45.809676 17490 annotated_data_layer.cpp:306]      Read time: 0.219 ms.
I1207 14:39:45.809682 17490 annotated_data_layer.cpp:307] Transform time: 8156.46 ms.
I1207 14:39:45.992069 17502 annotated_data_layer.cpp:305] Prefetch batch: 11379 ms.
I1207 14:39:45.992133 17502 annotated_data_layer.cpp:306]      Read time: 0.195 ms.
I1207 14:39:45.992141 17502 annotated_data_layer.cpp:307] Transform time: 11369.2 ms.
I1207 14:39:50.924690 17502 annotated_data_layer.cpp:305] Prefetch batch: 4898 ms.
I1207 14:39:50.924741 17502 annotated_data_layer.cpp:306]      Read time: 0.19 ms.
I1207 14:39:50.924746 17502 annotated_data_layer.cpp:307] Transform time: 4888.66 ms.
I1207 14:39:57.140404 17490 annotated_data_layer.cpp:305] Prefetch batch: 11292 ms.
I1207 14:39:57.140468 17490 annotated_data_layer.cpp:306]      Read time: 0.17 ms.
I1207 14:39:57.140476 17490 annotated_data_layer.cpp:307] Transform time: 11282 ms.
I1207 14:40:03.536177 17469 solver.cpp:243] Iteration 20, loss = 9.7219
I1207 14:40:03.536257 17469 solver.cpp:259]     Train net output #0: mbox_loss = 11.5929 (* 1 = 11.5929 loss)

you can see that the "Transform time" is too long, so my training is very slow. did anybody encounter this problem? could anybody can help me? looking for your anwsers! thank you very much. @weiliu89

**Comment by lubagov** [comment]
Data processing takes a lot of time. It looks very inefficient.
I see that all the data processing code is executed on one CPU core, and it is 100% loaded with JPEG image processing. In addition, inside src/caffe/layers/annotated_data_layer.cpp each image transformation, this->data_transformer_->(DistortImage, ExpandImage, CropImage, Transform) performs JPEG unpacking, then transformation, and compress JPEG back!!!

It seems strange, and very slow, if the input images are large (I use WILDER data, there are images of different sizes, from 600 to 1300 side).

As a result, the GPU is not loaded at all, but the CPU load 100% in one thread does nonsense.
